File: Concesionario/src/gestorAplicacion/activos/TransaccionVenta.java
package gestorAplicacion.activos;

import java.io.Serializable;
import java.util.ArrayList;

import gestorAplicacion.personal.*;

public class TransaccionVenta extends Transaccion implements Serializable{
	private static final long serialVersionUID = 1L;
	private Auto auto;
	private Vendedor vendedor;
	static ArrayList<TransaccionVenta> transaccionesven = new ArrayList<TransaccionVenta>();
	static ArrayList<Auto> autosV = new ArrayList<Auto>();
	static ArrayList<Vendedor> vend = new ArrayList<Vendedor>();
	static ArrayList<String> marcas = new ArrayList<>();
	
	public TransaccionVenta(String tipo, long ingreso, Cliente cliente, Auto auto, Vendedor vendedor, int transfer) {
		super(tipo, ingreso, cliente, transfer);
		this.auto=auto;
		this.vendedor=vendedor;
		transaccionesven.add(this);
		autosV.add(this.auto);
	}
	
	@Override
	public String info() {
		 String txt = String.format("Transacción #%08d: venta realizada por %s para el cliente %s por un total de $%d por el auto %s",
		            transfer, vendedor.getNombre(), cliente.getNombre(), auto.getPrecio(), auto.getModelo());
			
		
		return txt;
	}
	public static Cliente getClientePorCedula(long cedula){
		Transaccion finder = null;
		Cliente cli=null;
		
		for(Transaccion trans: transaccionesven) {
			if(trans.getClienteCed()== cedula) {
				finder = trans;
				cli=finder.getCliente();
				break;
			}
		}
		return cli;
	}
	public static Auto getTransaccionporCedula(long cedula) {
		Auto finder = null;
		for (TransaccionVenta trans: transaccionesven) {
			if(trans.getClienteCed()== cedula) {
				finder = trans.getAuto();
				break;
			}
		}
		return finder;
	}

	public void setAuto(Auto auto) {
	    this.auto = auto;
	}

	public Auto getAuto() {
	    return auto;
	}
	public void setVendedor(Vendedor vendedor) {
	    this.vendedor = vendedor;
	}

	public Vendedor getVendedor() {
	    return vendedor;
	}
	public static void agregarTransaccion(TransaccionVenta transaccion) {
	    transaccionesven.add(transaccion);
	}

	public static void eliminarTransaccion(TransaccionVenta transaccion) {
	    transaccionesven.remove(transaccion);
	}

	public static ArrayList<TransaccionVenta> getTransaccionesven() {
	    return transaccionesven;
	}
	
	// metodo para obtener los vendedores de las ventas realizadas y agregarlos a la lista "vendedores"
	public static ArrayList<Vendedor> vendedoresVentas(ArrayList<TransaccionVenta> ventas) {
	    for (TransaccionVenta transacc: ventas) {
	    	if (vend.contains(transacc.getVendedor())) {
	    	} else {
	    		vend.add(transacc.getVendedor());
	    	}
	    }
	    return vend;
	}
	
	public static ArrayList<String> marcasVentas(ArrayList<Auto> autosIniciales){
		for (Auto a: autosIniciales) {
			if (marcas.contains(a.getMarca())) {
			} else {
				marcas.add(a.getMarca());
			}
		}
		return marcas;
	}
	
	public static ArrayList<Auto> getAutosV(){
		return autosV;
	}
	
	public static long getIngresoPorAuto(Auto auto){
		for (TransaccionVenta transacc: TransaccionVenta.getTransaccionesven()) {
			if (transacc.getAuto()==auto) {
				return transacc.getIngreso();
			}
		}
		return 0;
	}
    public static ArrayList<Vendedor> getVend() {
        return vend;
    }

    public static void setVend(ArrayList<Vendedor> vendedores) {
        vend = vendedores;
    }

    public static ArrayList<String> getMarcas() {
        return marcas;
    }

    public static void setMarcas(ArrayList<String> listaMarcas) {
        marcas = listaMarcas;
    }
}

File: PYTHON/Concesionario/gestorAplicacion/Activos/transaccionventa.py
import sys
import os
ruta_activos = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'gestorAplicacion', 'Activos'))
ruta_personal = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'gestorAplicacion', 'Personal'))
ruta_gestor = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
sys.path.append(ruta_activos)
sys.path.append(ruta_personal)
sys.path.append(ruta_gestor)
from Personal import cliente, vendedor
from transaccion import Transaccion

class TransaccionVenta(Transaccion):
    transaccionesven = []
    autosV = []
    vend = []
    marcas = []

    def __init__(self, tipo, ingreso, cliente, auto, vendedor, transfer):
        super().__init__(tipo, ingreso, cliente, transfer)
        self.auto = auto
        self.vendedor = vendedor
        TransaccionVenta.transaccionesven.append(self)
        TransaccionVenta.autosV.append(self.auto)
    
    def info(self):
        txt = f"Transacción #{self.transfer:08d}: venta realizada por {self.vendedor.get_nombre()} para el cliente {self.cliente.get_nombre()} por un total de ${self.auto.get_precio()} por el auto {self.auto.get_modelo()}"
        return txt
    
    @staticmethod
    def get_cliente_por_cedula(cedula):
        finder = None
        cli = None

        for trans in TransaccionVenta.transaccionesven:
            if trans.getClienteCed() == cedula:
                finder = trans
                cli = finder.getCliente()
                break

        return cli
    
    @staticmethod
    def get_transaccion_por_cedula(cedula):
        finder = None
        for trans in TransaccionVenta.transaccionesven:
            if trans.get_cliente_ced() == cedula:
                finder = trans.get_auto()
                break
        return finder

    def set_auto(self, auto):
        self.auto = auto

    def get_auto(self):
        return self.auto
    
    def set_vendedor(self, vendedor):
        self.vendedor = vendedor
    
    def get_vendedor(self):
        return self.vendedor
    
    @staticmethod
    def agregar_transaccion(transaccion):
        TransaccionVenta.transaccionesven.append(transaccion)
    
    @staticmethod
    def eliminar_transaccion(transaccion):
        TransaccionVenta.transaccionesven.remove(transaccion)
    
    @staticmethod
    def get_transaccionesven():
        return TransaccionVenta.transaccionesven
    
    @staticmethod
    def vendedores_ventas(ventas):
        for transacc in ventas:
            if transacc.get_vendedor() not in TransaccionVenta.vend:
                TransaccionVenta.vend.append(transacc.get_vendedor())
        return TransaccionVenta.vend
    
    @staticmethod
    def marcas_ventas(autos_iniciales):
        for a in autos_iniciales:
            if a.get_marca() not in TransaccionVenta.marcas:
                TransaccionVenta.marcas.append(a.get_marca())
        return TransaccionVenta.marcas
    
    @staticmethod
    def get_autosV():
        return TransaccionVenta.autosV
    
    @staticmethod
    def get_ingreso_por_auto(auto):
        for transacc in TransaccionVenta.get_transaccionesven():
            if transacc.get_auto() == auto:
                return transacc.get_ingreso()
        return 0
    
    @staticmethod
    def get_vend():
        return TransaccionVenta.vend
    
    @staticmethod
    def set_vend(vendedores):
        TransaccionVenta.vend = vendedores
    
    @staticmethod
    def get_marcas():
        return TransaccionVenta.marcas
    
    @staticmethod
    def set_marcas(lista_marcas):
        TransaccionVenta.marcas = lista_marcas
